File: basic_dsa_four_hundred_sixty_seven.cpp
/**
 * Basic DSA - Separate the Digits in an Array (LeetCode 2553)
 * Return all decimal digits of nums[0], then nums[1], ... in order.
 */
#include <iostream>
#include <string>
#include <vector>

std::vector<int> separateDigits(const std::vector<int>& nums) {
    std::vector<int> ans;
    for (int x : nums) {
        for (char c : std::to_string(x))
            ans.push_back(c - '0');
    }
    return ans;
}

void printVec(const std::vector<int>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i)
            std::cout << ", ";
        std::cout << v[i];
    }
    std::cout << "]";
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Separate the Digits in an Array\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {13, 25, 2, 43};
    printVec(a1);
    std::cout << " -> ";
    printVec(separateDigits(a1));
    std::cout << "\n";

    std::vector<int> a2 = {7, 1, 3, 9};
    printVec(a2);
    std::cout << " -> ";
    printVec(separateDigits(a2));
    std::cout << "\n";
}

File: basic_dsa_four_hundred_sixty_seven.java
/**
 * Basic DSA Problems - One Fundamental Bit Manipulation Question
 * Simple and essential DSA problem for beginners
 */
public class basic_dsa_four_hundred_sixty_seven {

    // Problem 1: Single Number
    // Given a non-empty array of integers where every element appears twice except for one,
    // find that single one. You must use O(1) extra space (bitwise XOR approach).
    //
    // Example:
    // nums = [4,1,2,1,2] -> 4

    /**
     * XOR all elements; pairs cancel (a ^ a == 0)
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static Integer singleNumber(int[] nums) {
        if (nums == null || nums.length == 0) {
            return null;
        }

        int result = 0;
        for (int n : nums) {
            result ^= n;
        }
        return result;
    }

    private static String arrayToString(int[] arr) {
        if (arr == null) {
            return "null";
        }
        if (arr.length == 0) {
            return "[]";
        }
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < arr.length; i++) {
            if (i > 0) {
                sb.append(", ");
            }
            sb.append(arr[i]);
        }
        sb.append("]");
        return sb.toString();
    }

    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Single Number");
        System.out.println("=".repeat(60));

        int[] a1 = {4, 1, 2, 1, 2};
        System.out.println("nums: " + arrayToString(a1));
        System.out.println("Single: " + singleNumber(a1));

        int[] a2 = {2, 2, 1};
        System.out.println("\nnums: " + arrayToString(a2));
        System.out.println("Single: " + singleNumber(a2));

        int[] a3 = {1};
        System.out.println("\nnums: " + arrayToString(a3));
        System.out.println("Single: " + singleNumber(a3));

        int[] a4 = {};
        System.out.println("\nnums: " + arrayToString(a4));
        System.out.println("Single: " + singleNumber(a4));

        System.out.println("\nnums: null");
        System.out.println("Single: " + singleNumber(null));
    }
}
